File: 백준/Gold/14890. 경사로/경사로.py


n,l = map(int,input().split())

field = [list(map(int,input().split())) for _ in range(n)]
field_col = list(zip(*field))


answer = 0

check_row = [[False]*n for _ in range(n)]
check_col = [[False]*n for _ in range(n)]

def slope(field,check):
    global answer
    for i in range(n):
        temp = field[i][0]
        cnt = 101
        for j in range(1,n):
            if cnt < l:
                cnt+=1
                check[i][j-1]=True


            if abs(temp - field[i][j]) >1:
                break
            else:
                if temp>field[i][j]:
                    arr = set(field[i][j:j+l])
                    if len(field[i][j:j+l])!=l or len(arr)>1 or True in check[i][j:j+l]:
                        break
                    cnt=0

                elif temp<field[i][j]:
                    arr = set(field[i][j-l:j])
                    if len(field[i][j-l:j])!=l or len(arr) > 1 or True in check[i][j-l:j]:
                        break
            temp= field[i][j]
        else:
            answer+=1

slope(field,check_row)
slope(field_col,check_col)


print(answer)

File: 백준/Gold/14890. 경사로/경사로.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.StringTokenizer;

public class Main {
	static int N,L,answer=0;
	static int[][] map;
	static boolean[] row,col;
	
	public static void main(String[] args) throws Exception{
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		N = Integer.parseInt(st.nextToken());
		L = Integer.parseInt(st.nextToken());
		map = new int[N][N];
		
		for(int i=0; i<N; i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=0; j<N; j++) {
				map[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		
		//System.out.println("==================행");
		count(map);
		map = rotate(map);	
		//System.out.println("==================열");
		count(map);
		System.out.println(answer);
	}
	
	static void count(int[][] map) {
		for (int i = 0; i < N; i++) {
			//System.out.println("============i : "+i);
			boolean possible = true;
			boolean[] check = new boolean[N];
			int length = 1;
			int prev = map[i][0];
			for (int j = 1; j < N; j++) {
				//System.out.println("j "+j);
				
				if(Math.abs(map[i][j]-prev)>=2) {
					possible=false;
					break;
				}
				//오르막
				if(prev<map[i][j]) {
					//경사로 놓을 자리, 경사로가 이미 있는지 체크 
					//System.out.println("오르막");
					//System.out.println(Arrays.toString(check));
					if(j-L<0 || !isFlat(j-L, j, map[i]) || isSlope(j-L, j, check)) {
						//System.out.println("오르막 불가");
						possible =false;
						break;
					}
				}
				//내리막
				else if(prev > map[i][j]) {
					//경사로 놓을 자리 
					//System.out.println("내리막");
					
					if(j+L-1>=N || !isFlat(j, j+L, map[i])) {
						//System.out.println("내리막 불가");
						possible=false;
						break;
					}
					
					for(int k=j; k<j+L; k++) {
						check[k]=true;
					}
				}
				prev = map[i][j];
			}
			if(possible) {
				//System.out.println(i+"행 가능");
				answer++;
			}
		}
	}
	
	
	
	static int[][] rotate(int[][] map){
		int[][] arr = new int[N][N];
		
		for(int i=0; i<N; i++) {
			for(int j=0; j<N; j++) {
				arr[i][j] = map[j][i];
			}
		}
		
		return arr;
	}
	
	
	
	static boolean isSlope(int start, int end, boolean[] slope) {
		for (int i = start; i < end; i++) {
			if(slope[i]) {
				return true;
			}
		}
		return false;
	}
	
	
	
	static boolean isFlat(int start, int end, int[] map) {
		boolean flag = true;
		int height = -1;
		for(int i=start; i<end; i++) {
			if(height==-1) {
				height=map[i];
				continue;
			}
			if(height!=map[i]) {
				flag = false;
				break;
			}
		}
		return flag;
	}

}